web/components/contract-search.tsx

```diff
@@ -103,26 +91,73 @@ export function ContractSearch(props: {
     cardHideOptions,
     highlightOptions,
     headerClassName,
-    useQuerySortLocalStorage,
+    persistPrefix,
     useQuerySortUrlParams,
   } = props
 
-  const [numPages, setNumPages] = useState(1)
-  const [pages, setPages] = useState<Contract[][]>([])
-  const [showTime, setShowTime] = useState<ShowTime | undefined>()
+  const store = safeSessionStorage()
+  const persistAs = (name: string) => {
+    return persistPrefix ? { prefix: persistPrefix, name, store } : undefined
+  }
+
+  const [numPages, setNumPages] = usePersistentState(1, persistAs('numPages'))
+  const [pages, setPages] = usePersistentState<Contract[][]>(
+    [],
+    persistAs('pages')
+  )
+  const [showTime, setShowTime] = usePersistentState<ShowTime | null>(
+    null,
+    persistAs('showTime')
+  )
 
-  const searchParameters = useRef<SearchParameters | undefined>()
+  const searchParameters = useRef<SearchParameters | null>(null)
   const requestId = useRef(0)
 
+  useLayoutEffect(() => {
+    if (persistPrefix && store) {
+      const parameters = loadState(getKey(persistPrefix, 'parameters'), store)
+      const scrollY = loadState(getKey(persistPrefix, 'scrollY'), store)
+      if (parameters !== undefined) {
+        console.log('Restoring search parameters: ', parameters)
+        searchParameters.current = parameters as SearchParameters
+      }
+      if (scrollY !== undefined) {
+        console.log('Restoring scroll position: ', scrollY)
+        window.scrollTo(0, scrollY as number)
```

Comment:
So, we actually have another hook that is saving and restoring scroll position at the app level.

Check out `hooks/use-preserve-scroll.ts`. It would be great if this handled it for us and we didn't need to do anything special here. Otherwise, you might need to disable it somehow for this case.

Reply:
That looks promising, I didn't realize it existed. I don't know whether it will just work perfectly out of the box (probably it will?) but if we move forward on this approach I will definitely unify these in a way that makes sense.